<a href="https://colab.research.google.com/github/abhay43/ML_Code/blob/master/DL-MNIST_clothing_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import keras
from tensorflow.keras.layers import Dense
from keras import Sequential
import tensorflow as tf
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import Dropout 

Using TensorFlow backend.


In [0]:
data = keras.datasets.fashion_mnist

In [3]:
(X_train, Y_train), (X_test, Y_test) =  data.load_data()

4423680/4422102 [==============================] - 1s 0us/step


In [0]:
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train = X_train.reshape(len(X_train),784 )
X_test = X_test.reshape(len(X_test),784 )

In [0]:
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# Normalizing the input to be within the range [0,1]
X_train /= 255
#intensity of each pixel is divided by 255, the maximum intensity value
X_test /= 255

In [0]:

# One-hot representation of the labels.
Y_train = tf.keras.utils.to_categorical(Y_train, 10) 
Y_test = tf.keras.utils.to_categorical(Y_test, 10)

In [0]:
model_0 = tf.keras.Sequential()
hidden_layer = 10
model_0.add(Dense(hidden_layer, input_shape=(784,),name='dense_layer_1', activation='relu' ))
model_0.add(Dense(hidden_layer, name='dense_layer_2', activation='relu' ))
model_0.add(Dense(hidden_layer, name='dense_layer_3', activation='softmax' ))

In [42]:
Y_train.shape

(60000, 10)

In [0]:
model_0.compile(optimizer='SGD',  loss='categorical_crossentropy',metrics=['accuracy'])

In [46]:
training = model_0.fit(X_train, Y_train, batch_size=64, epochs=70, validation_split=0.2)

Epoch 1/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3500 - accuracy: 0.8758 - val_loss: 0.4029 - val_accuracy: 0.8618
Epoch 2/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3503 - accuracy: 0.8755 - val_loss: 0.4027 - val_accuracy: 0.8634
Epoch 3/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3480 - accuracy: 0.8777 - val_loss: 0.4193 - val_accuracy: 0.8562
Epoch 4/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3485 - accuracy: 0.8762 - val_loss: 0.4090 - val_accuracy: 0.8616
Epoch 5/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3489 - accuracy: 0.8769 - val_loss: 0.4211 - val_accuracy: 0.8496
Epoch 6/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3477 - accuracy: 0.8775 - val_loss: 0.4011 - val_accuracy: 0.8633
Epoch 7/70
750/750 [==============================] - 1s 2ms/step - loss: 0.3465 - accuracy: 0.8766 - val_loss: 0.4015 - val_accuracy: 0.8633
Epoch 

In [45]:
model_0.evaluate(X_test,Y_test)

313/313 [==============================] - 0s 1ms/step - loss: 0.4521 - accuracy: 0.8427


[0.45214423537254333, 0.8427000045776367]

**1) Do you get the same results if you run the Notebook multiple times without changing any parameters?** 
Answer - Yes it change the accuracy. 

In [0]:
model_1 = tf.keras.Sequential()
hidden_layer = 64
model_1.add(Dense(hidden_layer, input_shape = (784,), name='layer_1', activation='relu'))
model_1.add(Dense(hidden_layer, name='layer_2', activation='relu'))
model_1.add(Dense(10, name='layer_3', activation='softmax'))

In [0]:
model_1.compile(optimizer='SGD',  loss='categorical_crossentropy',metrics=['accuracy'])

In [65]:
training = model_1.fit(X_train, Y_train, batch_size=64, epochs=70, validation_split=0.2)

Epoch 1/70
750/750 [==============================] - 2s 2ms/step - loss: 1.0275 - accuracy: 0.6715 - val_loss: 0.6731 - val_accuracy: 0.7693
Epoch 2/70
750/750 [==============================] - 2s 2ms/step - loss: 0.6178 - accuracy: 0.7881 - val_loss: 0.5641 - val_accuracy: 0.8054
Epoch 3/70
750/750 [==============================] - 2s 2ms/step - loss: 0.5410 - accuracy: 0.8150 - val_loss: 0.5232 - val_accuracy: 0.8199
Epoch 4/70
750/750 [==============================] - 2s 2ms/step - loss: 0.5026 - accuracy: 0.8279 - val_loss: 0.4884 - val_accuracy: 0.8297
Epoch 5/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4791 - accuracy: 0.8345 - val_loss: 0.4848 - val_accuracy: 0.8275
Epoch 6/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4606 - accuracy: 0.8407 - val_loss: 0.4655 - val_accuracy: 0.8346
Epoch 7/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4470 - accuracy: 0.8451 - val_loss: 0.4446 - val_accuracy: 0.8450
Epoch 

In [66]:
model_1.evaluate(X_test, Y_test)

313/313 [==============================] - 0s 1ms/step - loss: 0.3640 - accuracy: 0.8732


[0.36401376128196716, 0.873199999332428]

**2) What is the effect of adding more neurons to each Conv2D layer?** Answer - It seems like it incress the accuracy if adding more neurons

In [0]:
model_2 = tf.keras.Sequential()
hidden_layer = 512
model_2.add(Dense(hidden_layer, input_shape = (784,), name='layer_1', activation='relu'))
model_2.add(Dropout(0.2))
model_2.add(Dense(hidden_layer, name='layer_2', activation='relu'))
model_2.add(Dropout(0.2))
model_2.add(Dense(10, name='layer_3', activation='softmax'))

In [10]:
model_2.compile(optimizer='SGD',  loss='categorical_crossentropy',metrics=['accuracy'])
training = model_2.fit(X_train, Y_train, batch_size=64, epochs=70, validation_split=0.2)

Epoch 1/70
750/750 [==============================] - 7s 10ms/step - loss: 0.9734 - accuracy: 0.6844 - val_loss: 0.6289 - val_accuracy: 0.7842
Epoch 2/70
750/750 [==============================] - 7s 9ms/step - loss: 0.6277 - accuracy: 0.7854 - val_loss: 0.5279 - val_accuracy: 0.8220
Epoch 3/70
750/750 [==============================] - 8s 10ms/step - loss: 0.5558 - accuracy: 0.8090 - val_loss: 0.4937 - val_accuracy: 0.8282
Epoch 4/70
750/750 [==============================] - 8s 10ms/step - loss: 0.5131 - accuracy: 0.8216 - val_loss: 0.4678 - val_accuracy: 0.8365
Epoch 5/70
750/750 [==============================] - 7s 10ms/step - loss: 0.4880 - accuracy: 0.8303 - val_loss: 0.4481 - val_accuracy: 0.8418
Epoch 6/70
750/750 [==============================] - 7s 10ms/step - loss: 0.4674 - accuracy: 0.8361 - val_loss: 0.4363 - val_accuracy: 0.8440
Epoch 7/70
750/750 [==============================] - 7s 10ms/step - loss: 0.4505 - accuracy: 0.8429 - val_loss: 0.4215 - val_accuracy: 0.8498


In [11]:
model_2.evaluate(X_test, Y_test)

313/313 [==============================] - 1s 3ms/step - loss: 0.3169 - accuracy: 0.8886


[0.31690990924835205, 0.8885999917984009]

**3) What happens if we manipulate the value of Dropout?** Answer- It incress the accuracy and reduce the overfitting

In [13]:
model_3 = tf.keras.Sequential()
hidden_layer = 32
model_3.add(Dense(hidden_layer, input_shape = (784,), name='layer_1', activation='relu'))
model_3.add(Dense(hidden_layer, name='layer_2', activation='relu'))
model_3.add(Dense(hidden_layer, name='layer_3', activation='relu'))
model_3.add(Dense(hidden_layer, name='layer_4', activation='relu'))
model_3.add(Dense(10, name='layer_5', activation='softmax'))

model_3.compile(optimizer='SGD',  loss='categorical_crossentropy',metrics=['accuracy'] )
training = model_3.fit(X_train, Y_train, batch_size=64, epochs=70, validation_split=0.2)

Epoch 1/70
750/750 [==============================] - 2s 2ms/step - loss: 1.2558 - accuracy: 0.5606 - val_loss: 0.7103 - val_accuracy: 0.7448
Epoch 2/70
750/750 [==============================] - 1s 2ms/step - loss: 0.6478 - accuracy: 0.7694 - val_loss: 0.6474 - val_accuracy: 0.7706
Epoch 3/70
750/750 [==============================] - 1s 2ms/step - loss: 0.5606 - accuracy: 0.8011 - val_loss: 0.5486 - val_accuracy: 0.7998
Epoch 4/70
750/750 [==============================] - 1s 2ms/step - loss: 0.5129 - accuracy: 0.8191 - val_loss: 0.5058 - val_accuracy: 0.8217
Epoch 5/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4833 - accuracy: 0.8298 - val_loss: 0.4856 - val_accuracy: 0.8313
Epoch 6/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4575 - accuracy: 0.8388 - val_loss: 0.4548 - val_accuracy: 0.8378
Epoch 7/70
750/750 [==============================] - 2s 2ms/step - loss: 0.4407 - accuracy: 0.8450 - val_loss: 0.4412 - val_accuracy: 0.8449
Epoch 

In [14]:
model_3.evaluate(X_test,Y_test)

313/313 [==============================] - 0s 1ms/step - loss: 0.3851 - accuracy: 0.8674


[0.38507112860679626, 0.8673999905586243]

**4) What is the effect of adding more hiddenlayers to the network?** Answer - It does not add any value in this case if I added more layer in the model. But generally we can improves the performance 

**5) Does manipulating the learning rate affect the model? Justify your answer.** Answer - Learning rate is used to control the greedy/ non-greedy strategy.It can be positively effect or also it can negatively effect the model. If large learning rate is given then a small change in the value may effect the performance.  

**6) What is the best parameter configuration for this project?** Answer - model_2 is having 88.86% accuracy. So the parameters that is used there is the most optimum